pkg/migrator/migrator.go

```diff
@@ -0,0 +1,264 @@
+// operator installation as well the stale datastore CSCs when the cluster is
+// migrating from Openshift 4.1.z to Openshift 4.2.z
+func (m *migrator) doMigrate(operatorNamespace string) []error {
+	installedCscs, errors := m.updateSubscriptions()
+	errors = append(errors, m.deleteInstalledCscs(installedCscs, operatorNamespace)...)
+	errors = append(errors, m.deleteDatastoreCscs(operatorNamespace)...)
+	return errors
+}
+
+// Migrate is a wrapper for doMigrate in order to retry the migration
+// retryNumber times before reporting an error
+func (m *migrator) Migrate(operatorNamespace string) error {
+	for i := retryNumber; i > 0; i-- {
+		err := m.doMigrate(operatorNamespace)
+		if err == nil {
+			return nil
+		}
+	}
+	return fmt.Errorf("[migration] Migration Failed to Complete")
+}
+
+// updateSubscriptions updates the existing Subscriptions'
+// spec.source and spec.sourcenamespace fields. Existing
+// Subscriptions referenced installed CatalogSources, which
+// are updated to reference datastore CatalogSources
+// instead.
+func (m *migrator) updateSubscriptions() ([]types.NamespacedName, []error) {
+	var installedCscs []types.NamespacedName
+	var errors []error
+	options := &client.ListOptions{}
+	options.SetLabelSelector(fmt.Sprintf(builders.CscOwnerNameLabel))
+
+	subscriptions := &olm.SubscriptionList{}
+	// Get the list of existing Subscriptions that have the label "csc-owner-name"
+	// which is a label added to Subscriptions in the 4.1 UI
+	err := m.client.List(context.TODO(), options, subscriptions)
+	if err != nil {
+		msg := fmt.Sprintf("[migration] Client error: %s", err.Error())
+		errors = append(errors, fmt.Errorf(msg))
+		m.logger.Errorf(msg)
+		return []types.NamespacedName{}, errors
+	}
+	for _, subscription := range subscriptions.Items {
+		if builders.HasOwnerLabels(subscription.GetLabels(), v2.CatalogSourceConfigKind) {
+			// Try to infer the datastore CatalogSource from the Subscription
+			datastoreCs, err := findCatalogSource(&subscription, m.client)
+			if k8s_errors.IsNotFound(err) {
+				// Infer the CatalogSource from the OperatorSource that has the package
+				datastoreCs, err = findCsFromOpsrc(&subscription, m.client)
+				if err != nil {
+					msg := fmt.Sprintf("[migration] Could not infer datastore CatalogSource for Subscription %s: %v", subscription.GetName(), err)
+					errors = append(errors, fmt.Errorf(msg))
+					m.logger.Errorf(msg)
+					continue
+				}
+			}
+			// Update the Subscription to reference the datastore CatalogSource
+			subscription.Spec.CatalogSource = datastoreCs.GetName()
+			subscription.Spec.CatalogSourceNamespace = datastoreCs.GetNamespace()
+			// Remove the owner labels from the subscription
+			subscription.Labels = labelsutil.CloneAndRemoveLabel(
+				labelsutil.CloneAndRemoveLabel(subscription.GetLabels(), builders.CscOwnerNameLabel),
+				builders.CscOwnerNamespaceLabel)
+			err = m.client.Update(context.TODO(), &subscription)
```

Comment:
Any reason not to patch? OLM also reads and writes Subscriptions, so if we can reduce the chance of conflicts that would be ideal.

Reply:
I see - I didn't think of that, thanks for pointing it out!

Reply:
controller-runtime and by extension the SDK client does not provide `patch`. Please see https://github.com/operator-framework/operator-sdk/issues/1077. We would have to move using the raw REST client to do that. Might too much of a change at this point.

Reply:
Made a note to bring this up next sync - I think patch is pretty important for reducing collisions.